### PR TITLE
SALTO-5995 init Jamf adapter

### PR DESCRIPTION
**Pull request by omrifarkash**
First working version of Jamf adapter

---

---
_Release Notes_: 
_Jamf Adapter_:
- first version released

---
_User Notifications_: 
_Jamf Adapter_:
- first version released

### CODE REVIEW

File: packages/jamf-adapter/src/definitions/requests/clients.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            omitBody: true,
+          },
+        },
+      },
+    },
+    classicApi: {
+      httpClient: clients.main,
+      endpoints: {
+        default: {
+          get: {
```
Comment (omrifarkash) [inline packages/jamf-adapter/src/definitions/requests/clients.ts:43]: https://salto-io.atlassian.net/browse/SALTO-6128
add pagination function to this client

File: packages/adapter-components/src/definitions/system/shared/transformation.ts [hunk 2]
```diff
@@ -66,7 +66,7 @@ export type AdjustFunction<TContext = ContextParams, TSourceVal = unknown, TTarg
  *   if single is true, the first result of the transformation will be returned from the resulting array
  *   (and undefined will be returned if the array is empty). if single is false, the result will be returned as-is.
  */
-export type TransformDefinition<TContext = ContextParams, TTargetVal = Values> = {
+export type TransformDefinition<TContext = ContextParams, TTargetVal = Values | Value> = {
```
Comment (omrifarkash) [inline packages/adapter-components/src/definitions/system/shared/transformation.ts:69]: In Jamf classic api we need to send an XML to the api, XML is represented as string and Values is record<string, any>

File: packages/jamf-adapter/src/definitions/deploy/deploy.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  },
+  dependencies: [
+    // {
+    //   first: { type: 'dynamic_content_item', action: 'add' },
+    //   second: { type: 'dynamic_content_item_variant', action: 'add' },
+    // },
+    // {
+    //   first: { type: 'dynamic_content_item', action: 'remove' },
+    //   second: { type: 'dynamic_content_item_variant', action: 'remove' },
+    // },
```
Comment (shir-reifenberg) [inline packages/jamf-adapter/src/definitions/deploy/deploy.ts:110]: should we remove ?

**Review by shir-reifenberg — COMMENTED** [review]
Looks great!!! added some comments

File: packages/jamf-adapter/src/definitions/fetch/transforms/class.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+/*
+ * Adjust function to convert site object to site id to make reference
+ */
+
+export const adjust: definitions.AdjustFunction = ({ value }) => {
+  if (!values.isPlainRecord(value)) {
+    throw new Error('Expected value to be a record')
+  }
+  const siteId = _.get(value, 'site.id')
```
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/fetch/transforms/class.ts:28]: I am ok with putting `undefined` if site.id is not exist

File: packages/jamf-adapter/src/definitions/fetch/transforms/policy.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  const siteId = _.get(value, 'general.site.id')
+  _.set(value, 'general.site', siteId === -1 ? _.get(value, 'general.site.name') : siteId)
+}
+
+/*
+ * Convert category object to category id to make reference
+ */
+const adjustCategoryObjectToCategoryId = (value: Record<string, unknown>): void => {
+  const categoryId = _.get(value, 'general.category.id')
+  _.set(value, 'general.category', categoryId === -1 ? _.get(value, 'general.category.name') : categoryId)
```
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/fetch/transforms/policy.ts:33]: Correct, but if there is not category for policy, the service returns `category: { id: -1 }`

File: packages/jamf-adapter/src/definitions/requests/pagination.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ */
+import { definitions } from '@salto-io/adapter-components'
+import { logger } from '@salto-io/logging'
+import _ from 'lodash'
+import { ClientOptions, PaginationOptions } from '../types'
+
+const log = logger(module)
+
+export const PAGINATION: Record<PaginationOptions, definitions.PaginationDefinitions<ClientOptions>> = {
+  increasePageUntilEmpty: {
```
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/requests/pagination.ts:24]: Not the same behavior :/

File: packages/jamf-adapter/src/definitions/fetch/fetch.ts [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+  },
+  updated_at: {
+    hide: true,
+  },
+  created_by_id: {
+    hide: true,
+  },
+  updated_by_id: {
+    hide: true,
+  },
```
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/fetch/fetch.ts:45]: These come from placeholder adapter like this

File: packages/jamf-adapter/src/definitions/fetch/fetch.ts [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+    },
+  },
+  [SITE_TYPE_NAME]: {
+    requests: [
+      {
+        endpoint: {
+          path: '/api/v1/sites',
+          // this is not a classic API, but no pagination function like classic api
+          client: 'classicApi',
+        },
```
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/fetch/fetch.ts:205]: Jamf has two types of APIs `classic` and `pro`

This endpoint belongs to `pro` api but behaves like `classic` api pagination wise, So I am using `classicApi` client even though it is not an classic api endpoint

File: packages/jamf-adapter/src/definitions/fetch/fetch.ts [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+        transformation: {
+          root: 'policy',
+          adjust: transforms.adjustPolicy,
+        },
+      },
+    ],
+    resource: {
+      directFetch: true,
+      context: {
+        dependsOn: {
```
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/fetch/fetch.ts:323]: We created instances only for `x` and not `x_minimal` you can see that `x_minimal` is not defined as `topLevel`.

We have a query that brings all x ids (this is `x_minimal`), then, per id we need to usr `x` definition to fetch its data.


(Jamf classic api is sucks)

File: packages/jamf-adapter/src/definitions/fetch/fetch.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/jamf-adapter/src/definitions/fetch/fetch.ts:46]: so can you please remove if it's not actual fields in pager duty / convert to `omit` if there's no justification for keeping this as hidden ?

File: packages/jamf-adapter/src/definitions/fetch/fetch.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/jamf-adapter/src/definitions/fetch/fetch.ts:206]: got it. I think you can adjust pagination function for endpoints in `clients` configuration, like this - 
```
          [SITE_TYPE_NAME]: {
            get: {
              pagination: 'myPaginationFunc',
            }
          },
```

If possible, I think it's best to keep it in its actual client, cause all the rate limit calculation are done per client.

File: packages/jamf-adapter/src/definitions/fetch/fetch.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/jamf-adapter/src/definitions/fetch/fetch.ts:324]: got it, thanks 🙏 
maybe a long shot - but it's worth checking if there's an additional query param to pass to the "minimal" endpoint that would return full results (you can also check what the UI asks for)

File: packages/jamf-adapter/src/definitions/requests/pagination.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/jamf-adapter/src/definitions/requests/pagination.ts:24]: I copied the wrong function name 🤦‍♀️ 
I thought we can use `pageOffsetPagination` with `dataField` as `results` and `paginationField` as `page`

**Review by shir-reifenberg — APPROVED** [review]
💯 💯 looks great
Added a few tiny things, I made a mistake with the pagination earlier but I think we alrady have this logic and can re-use

File: packages/jamf-adapter/src/definitions/requests/pagination.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/requests/pagination.ts:24]: I think there is a bug in `getItems` function. 
Anyway it is not working for me now

File: packages/jamf-adapter/src/definitions/fetch/fetch.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/fetch/fetch.ts:206]: The problem is `myPaginationFunc` is not exist yet
And `pagination: undefined` is getting override by the default definitions (which is `increasePageUntilEmpty` in that case)

File: packages/jamf-adapter/src/definitions/fetch/fetch.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/fetch/fetch.ts:324]: UI shows a list of `x` with the minimal query. when clicking on an item, it is querying for full data per the id

**Pull request merged**